specs/validator/0_beacon-chain-validator.md

```diff
@@ -127,13 +127,61 @@ Once a validator is activated, the validator is assigned [responsibilities](#bea
 
 *Note*: There is a maximum validator churn per finalized epoch so the delay until activation is variable depending upon finality, total active validator balance, and the number of validators in the queue to be activated.
 
+## Validator assignments
+
+A validator can get committee assignments for a given epoch using the following helper via `get_committee_assignment(state, epoch, validator_index)` where `epoch <= next_epoch`.
+
+```python
+def get_committee_assignment(
+        state: BeaconState,
+        epoch: Epoch,
+        validator_index: ValidatorIndex) -> Tuple[List[ValidatorIndex], Shard, Slot]:
+    """
+    Return the committee assignment in the ``epoch`` for ``validator_index``.
+    ``assignment`` returned is a tuple of the following form:
+        * ``assignment[0]`` is the list of validators in the committee
+        * ``assignment[1]`` is the shard to which the committee is assigned
+        * ``assignment[2]`` is the slot at which the committee is assigned
+    """
+    next_epoch = get_current_epoch(state) + 1
+    assert epoch <= next_epoch
+
+    committees_per_slot = get_epoch_committee_count(state, epoch) // SLOTS_PER_EPOCH
+    epoch_start_slot = get_epoch_start_slot(epoch)
+    for slot in range(epoch_start_slot, epoch_start_slot + SLOTS_PER_EPOCH)
+        slot_start_shard = get_epoch_start_shard(state, epoch) + committees_per_slot * (slot % SLOTS_PER_EPOCH)
+        for i in range(committees_per_slot):
+            shard = (slot_start_shard + i) % SHARD_COUNT
+            committee = get_crosslink_committee(state, epoch, shard)
+            if validator_index in committee:
+                return committee, shard, slot
+```
+
+A validator can use the following function to see if they are supposed to propose during their assigned committee slot. This function can only be run during the slot in question. Proposer selection is only stable within the context of the current epoch.
+
+```python
+def is_proposer(state: BeaconState,
+                validator_index: ValidatorIndex) -> bool:
+    return get_beacon_proposer_index(state) == validator_index
+```
+
+*Note*: To see if a validator is assigned to proposer during the slot, the validator must run an empty slot transition from the previous state to the current slot using `process_slots(state, current_slot)`.
```

Comment:
Transitioning to the previous slot is no longer required (see other comment).

Reply:
ah, right. balance _only_ changes at epoch's right?

Reply:
gotcha. `effective_balance` is only adjusted at epoch boundaries. A few things still change `balance` but it isn't cleaned up against `effective_balance` until the boundaries.

This is a good invariant to note. from `state_A -> state_b` all `effective_balances` are unchanged unless `state_A -> state_B` is an epoch change.